### PR TITLE
refactor[venom]: refactor venom operand classes

### PR DESCRIPTION
**Pull request by harkal**
### Commit message
```
the original `IRValue`, `IRVariable`, `IRLabel` and `IROperand`
hierarchy got convoluted after several refactors. this commit refactors
the hierarchy so that the relationship between the classes is clearer.
```
### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Comment by charles-cooper** [comment]
seems fine. i think we should have explicit `__eq__()` implementations, and as a style thing i personally prefer explicit isinstance checks (friendlier to mypy) instead of the utility functions, but it OK either way tbh.

**Comment by harkal** [comment]
> seems fine. i think we should have explicit `__eq__()` implementations, and as a style thing i personally prefer explicit isinstance checks (friendlier to mypy) instead of the utility functions, but it is OK either way tbh.

I reverted the utility functions, `isinstance` is the way everywhere now.  Code check complaining about `__eq__()` is immaterial, the methods will be exactly the same in all subclasses.

File: vyper/venom/basicblock.py [hunk 1]
```diff
@@ -174,6 +174,8 @@ def __eq__(self, other):
         return super().__eq__(other)
 
     def __hash__(self):
+        # https://docs.python.org/3/reference/datamodel.html#object.__hash__
+        # https://cdn3.emoji.gg/emojis/2845-ptsd.png
```
Comment (charles-cooper) [inline vyper/venom/basicblock.py:178]:
```suggestion
        # __hash__ is required when __eq__ is overridden --
        # https://docs.python.org/3/reference/datamodel.html#object.__hash__
```

**Review by charles-cooper — APPROVED** [review]

File: vyper/venom/basicblock.py [hunk 2]
```diff
@@ -192,16 +169,14 @@ def __init__(self, value: str, is_symbol: bool = False) -> None:
         self.value = value
         self.is_symbol = is_symbol
 
-    def __hash__(self) -> int:
-        return hash(self.value)
+    def __eq__(self, other):
+        # is_symbol does not participate in equality
```
Comment (charles-cooper) [inline vyper/venom/basicblock.py:173]:
```suggestion
        # no need for is_symbol to participate in equality
```

**Pull request merged**